src/field.rs

```diff
@@ -31,6 +44,28 @@ pub struct Fp31(<Self as Field>::Integer);
 impl Field for Fp31 {
     type Integer = u8;
     const PRIME: Self::Integer = 31;
+    const ZERO: Self = Fp31(0);
+    const ONE: Self = Fp31(1);
+
+    fn invert(&self) -> Self {
+        debug_assert!(
+            self != &Self::ZERO,
+            "Multiplicative inverse is not defined for Fp31(0)"
+        );
+        self.pow(Self::PRIME - 2)
+    }
+
+    fn pow(&self, exp: Self::Integer) -> Self {
+        let mut t = Self::ONE;
+        for i in (0..Self::Integer::BITS - exp.leading_zeros()).rev() {
+            t *= t;
+            if (exp >> i) & 1 != 0 {
+                t *= *self;
+            }
+        }
+
+        t
+    }
```

Comment:
Perhaps add a comment to explain what you're doing here? I'm not good at following bitwise logic like this.

Reply:
ack